Zypper package source property addition

manick-vel-11 opened the pull request:
## Description
 We have added the functionality to zypper package for supporting source property.

A sample recipe is mentioned below.

zypper_package 'splunk' do
  action :install
  source '/home/ec2-user/splunk-8.2.1-ddff1c41e5cf-linux-2.6-x86_64.rpm'
  version "8.2.1"
end

This recipe will pick the package from the source path and install it. Before this change source was not supported for zypper package provider.  

## Related Issue
 https://github.com/chef/chef/issues/11638 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] I have read the **CONTRIBUTING** document.
- [x ] I have run the pre-merge tests locally and they pass.
- [x ] I have updated the documentation accordingly.
- [x ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

manick-vel-11 commented:
Could anyone please review this PR

btm commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -70,7 +71,34 @@ def get_current_versions
+          scheme = URI.split(str).first
+          return false unless scheme
+
+          %w{http https ftp file}.include?(scheme.downcase)
```
How did you determine this list? There's a long list [here](https://en.opensuse.org/openSUSE:Libzypp_URIs), not all of which might work for us because we'll have to deal with passing them to rpm in this implementation but I'm curious why these 4.

btm commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -70,7 +71,34 @@ def get_current_versions
 
         def candidate_version
-          @candidate_version ||= package_name_array.each_with_index.map { |pkg, i| available_version(i) }
+          if source_files_exist?
```
This is presuming the source property is a string, which it is for the `package` resource, but for the `dpkg_package` resource that you're modeling after it can be an array.

btm commented on spec/unit/provider/package/zypper_spec.rb:
```diff
@@ -32,6 +32,8 @@
 
   let(:status) { double(stdout: "\n", exitstatus: 0) }
 
+  let(:source) { "/tmp/wget_1.11.4-1ubuntu1_amd64.deb" }
```
we shouldn't be using "deb" package names in the zypper specs, since it uses rpms.

btm commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -70,7 +71,34 @@ def get_current_versions
         def candidate_version
-          @candidate_version ||= package_name_array.each_with_index.map { |pkg, i| available_version(i) }
+          if source_files_exist?
+            unless uri_scheme?(new_resource.source) || ::File.exist?(new_resource.source)
```
we shouldn't actually ever get here with a uri, because `source_files_exist?` would be false.

btm reviewed (changes requested):
This is missing the assertion for `source_files_exist?` from `Package::Dpkg`, so if you specify a source but it doesn't exist, we silently continue with the other behavior of looking for a package based on the resource name.

The URI code doesn't have any unit tests, we should have something there.

manick-vel-11 commented:
I have addressed the comments, please have a look at them. Regarding URI scheme (https://github.com/chef/chef/pull/12182#discussion_r755511857), since it is not necessary I have removed those parts and tested the changes in EC2 suse instance.

Test recipe seems to be running fine with source and without source.

Regarding the multipackage support in zypper, the same exception  **Chef::Exceptions::InvalidResourceSpecification: You may not specify both multipackage and source** is being thrown similar to apt package. Please suggest how to go about handling situation wherein both multipackage and source at the same time. Attaching screenshot for reference.
![zypper_multipackage](https://user-images.githubusercontent.com/83069519/150795072-9e52a22f-0905-43fc-bd79-72c9086ecd76.png)

lamont-granquist commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -70,7 +71,20 @@ def get_current_versions
 
         def candidate_version
-          @candidate_version ||= package_name_array.each_with_index.map { |pkg, i| available_version(i) }
+          if source_files_exist?
```
this helper should really go away and just be `if new_resource.source` checks since if that value is not-nil then you're doing source installs.

lamont-granquist commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -70,7 +71,20 @@ def get_current_versions
+                new_resource.version($2)
+                @candidate_version = $2
+              end
+            end
```
the shell_out and its parsing here should probably be extracted out to its own method, see `resolve_source_to_version_obj` in the dnf or yum providers (although you don't have a version obj in this case so probably `resolve_source_to_version`.

lamont-granquist replied:
(in general the dnf provider is a reasonably good template for how the code change should look)

lamont-granquist commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -127,6 +141,10 @@ def installed_version(index)
 
+        def source_files_exist?
+          resolved_source_array.all? { |s| s && ::File.exist?(s) }
+        end
```
this helper should be deleted in favor of `if new_resource.source`

lamont-granquist commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -139,7 +157,7 @@ def zypper_version
+          zipped_names = source_files_exist? ? [new_resource.source] : zip(names, versions)
           if zypper_version < 1.0
             shell_out!("zypper", global_options, gpg_checks, command, *options, "-y", names)
           else
```
there should just be one big conditional here which has `if new_resource.resource` followed by the zypper command to install all the sources (doesn't need to be flattened or anything as well since `shell_out!` will handle flattening arrays).  then the `else` side of the condition should be all the existing code.

lamont-granquist commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -139,7 +157,7 @@ def zypper_version
 
         def zypper_package(command, global_options, *options, names, versions)
-          zipped_names = zip(names, versions)
+          zipped_names = source_files_exist? ? [new_resource.source] : zip(names, versions)
```
i think you can drop array-ifying the source property, but then you need a check to fix the zypper_version < 1.0 case:

```
          zipped_names = new_resource.source ? new_resource.source : zip(names, versions)
          names = zipped_names if new_resource.source
```

lamont-granquist commented:
I'd also expect to see this check added or something very similar to it:

https://github.com/chef/chef/blob/main/lib/chef/provider/package/dnf.rb#L85-L93

manick-vel-11 commented:
I have replaced `source_file_exists?` related changes with new_resource.source. 

Regarding resolve_source_to_version, took reference from `dnf`, `yum` packages and added version class under `zypper` too as suggested.

Test recipe seems to be working fine with new changes. Please take a look

lamont-granquist commented on lib/chef/provider/package/zypper.rb:
```diff
@@ -70,7 +72,22 @@ def get_current_versions
-          @candidate_version ||= package_name_array.each_with_index.map { |pkg, i| available_version(i) }
+          if new_resource.source
+            logger.trace("#{new_resource} checking rpm status")
+            resolve_source_to_version
```
this won't work because you're passing back one of the Version objects into the candidate_version.

if you look at the yum package provider this code goes into the `available_version` helper and then that is rendered into an string by the loop which is in `candidate_version` there.  if we go down the road of adding the Version object then that pattern needs to be adhered to.

this also calls out the fact that this code needs functional tests, because this code as it stands is pretty certainly broken but all the unit tests pass.  so we need code in spec/functional/resource/zypper_package_spec.rb which does a real install of a source package (i think there's examples in the yum/dnf functional tests a well, and the zypper func tests are written similarly so it should be easy enough to copy and get a working functional test).

lamont-granquist commented:
This is blowing up on the specific functional test because the zypper_package provider doesn't have a cache, so doesn't have flush_cache:

```
 
  1) Chef::Resource::ZypperPackage installing a package install package with source argument
     Failure/Error: Chef::Resource::ZypperPackage.new("shouldnt-matter", run_context).run_action(:flush_cache)
 
     NoMethodError:
       zypper_package[shouldnt-matter] (dynamically defined) had an error: NoMethodError: undefined method `action_flush_cache' for 
[...]
```

lamont-granquist commented on spec/unit/provider/package/zypper_spec.rb:
```diff
@@ -163,6 +165,15 @@ def shell_out_expectation!(*command, **options)
+      shell_out_expectation!(
+        "zypper", "--non-interactive", "install", "--auto-agree-with-licenses", "--oldpackage", "/tmp/wget_1.11.4-1ubuntu1_amd64.rpm"
+      )
+      provider.install_package(["emacs"], ["1.0"])
```
these tests should have the correct behavior from the superclass which i believe is:

```      
provider.install_package(["wget"], ["1.11.4-1ubuntu1_amd64"])
```

or something close to that.

this is also insufficient for testing the actual cooperation between the superclass and the subclass and making sure that the subclass is filling out the `candidate_version` and `installed_version` correctly along with the correct `package_name_array` so that the call from the superclass into `install_package` has the correct information.

that needs to be done correctly, even if the install_package routine goes and trusts the source, because the idempotent behavior is handled by the superclass and that is what needs to be correct.

johnmccrae commented:
Hi can you please rebase this off main and push it again? We fixed the problems with Windows integration tests

saghoshprogress commented:
@PrajaktaPurohit can you please take a look and merge the same.

PrajaktaPurohit commented on spec/unit/provider/package/zypper_spec.rb:
```diff
@@ -200,6 +211,16 @@ def shell_out_expectation!(*command, **options)
+      )
+      provider.install_package(["wget"], ["1.11.4-1ubuntu1_amd64"])
+    end
+
```
Could there be some tests here to show what happens:
1. when the version is given but does not match the source package
2. when the source file is missing
3. version is given

jaymzh replied:
agreed

pratixha replied:
@PrajaktaPurohit  I have added functional and unit test cases for missing source path. 
Ready for review.

PrajaktaPurohit reviewed (changes requested):
Request for some additional functional test case scenarios.

PrajaktaPurohit commented on spec/unit/provider/package/zypper_spec.rb:
```diff
@@ -163,6 +165,15 @@ def shell_out_expectation!(*command, **options)
+        "zypper", "--non-interactive", "install", "--auto-agree-with-licenses", "--oldpackage", "/tmp/wget_1.11.4-1ubuntu1_amd64.rpm"
+      )
+      provider.install_package(["wget"], ["1.11.4-1ubuntu1_amd64"])
+    end
```
Could there be a test here showing what happens when:
1. when the version is given but does not match the source package

jaymzh replied:
I vaguely seem to remember us discussing this a few weeks back before Lamont left, and we decided it was undefined behavior. Currently, the `version` property is only used to determine if an action is needed (doesn't it match the installed version), and then if it doesn't, we install whatever the source is. If the source isn't the right version, this of course won't be idempotent and it'll run again on the next run. But AFAICT, we've never checked for this in any provider and the behavior is undefined.

pratixha replied:
@PrajaktaPurohit  Version attributed is overlooked if source is provided in any recipe and if the functionality has to change then it should be raised as a separate issue according to me. This is how its working for all the other package handlers like dpkg, rpm .. I have tested the same.

jaymzh reviewed (approved):
Assuming tests pass, this seems good.

PrajaktaPurohit reviewed (approved):
This will need a rebase with the chef main branch to get the tests passing. Thanks!
